platform/SharedPtr.h

```diff
@@ -0,0 +1,275 @@
+  *     struct MyStruct { int a; };
+  * 
+  *     // Create shared pointer
+  *     SharedPtr<MyStruct> ptr( new MyStruct );
+  * 
+  *     // Increase reference count
+  *     SharedPtr<MyStruct> ptr2( ptr );
+  * 
+  *     ptr = NULL; // Reference to the struct instance is still held by ptr2
+  * 
+  *     ptr2 = NULL; // The raw pointer is freed
+  * }
+  * @endcode
+  * 
+  * 
+  * It is similar to the std::shared_ptr class introduced in C++11, 
+  * however this is not a compatible implementation (no weak pointer, no make_shared, no custom deleters, etc.)
+  *
+  * Usage: SharedPtr<Class> ptr(new Class())
+  *
+  * When ptr is passed around by value the copy constructor and
+  * destructor manages the reference count of the raw pointer.
+  * If the counter reaches zero, delete is called on the raw pointer.
+  *
+  * To avoid loops, "weak" references should be used by calling the original
+  * pointer directly through ptr.get().
+  */
+
+template <class T>
+class SharedPtr {
+public:
+    /**
+     * @brief Create empty SharedPtr not pointing to anything.
+     * @details Used for variable declaration.
+     */
+    SharedPtr(): _ptr(NULL), _counter(NULL) {
+    }
+
+    /**
+     * @brief Create new SharedPtr
+     * @param ptr Pointer to take control over
+     */
+    SharedPtr(T* ptr): _ptr(ptr), _counter(NULL) {
+        // allocate counter on the heap so it can be shared
+        if(_ptr != NULL) {
+            _counter = (uint32_t*) malloc(sizeof(uint32_t));
+            *_counter = 1;
+        }
+    }
+
+    /**
+     * @brief Destructor.
+     * @details Decrement reference counter and delete object if no longer pointed to.
+     */
+    ~SharedPtr() {
+        decrement_counter();
+    }
+
+    /**
+     * @brief Copy constructor.
+     * @details Create new SharedPtr from other SharedPtr by
+     *          copying pointer to original object and pointer to counter.
+     * @param source Object being copied from.
+     */
+    SharedPtr(const SharedPtr& source): _ptr(source._ptr), _counter(source._counter) {
+        // increment reference counter
+        if (_ptr != NULL) {
+            core_util_critical_section_enter();
+            (*_counter)++;
+            core_util_critical_section_exit();
+        }
+    }
+
+    /**
+     * @brief Assignment operator.
+     * @details Cleanup previous reference and assign new pointer and counter.
+     * @param source Object being assigned from.
+     * @return Object being assigned.
+     */
+    SharedPtr operator=(const SharedPtr& source) {
+        if (this != &source) {
+            // clean up by decrementing counter
+            decrement_counter();
+
+            // assign new values
+            _ptr = source.get();
+            _counter = source.get_counter();
+
+            // increment new counter
+            if (_ptr != NULL) {
+                core_util_critical_section_enter();
+                (*_counter)++;
+                core_util_critical_section_exit();
```

Comment:
Hmm, is [core_util_atomic_incr_u32](https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_critical.h#L342) applicable here?

Not a blocker, adopting atomic increments could be an optimization for later